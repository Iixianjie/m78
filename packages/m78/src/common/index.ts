export * from "./types/index.js";

export * from "./utils/index.js";

export * from "./consts/index.js";

export * from "./unit/index.js";
