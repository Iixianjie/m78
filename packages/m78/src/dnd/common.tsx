import { createContext, useMemo } from "react";
import {
  _DNDMapEntry,
  _GroupState,
  DNDEnableInfos,
  DNDNode,
  DNDPosition,
  DNDProps,
  DNDStatus,
  _Context,
  _PendingItem,
  _LevelContext,
  _AutoScrollCtx,
} from "./types.js";
import {
  AnyObject,
  Bound,
  hasScroll,
  isBoolean,
  isFunction,
  isObject,
  raf,
  TupleNumber,
} from "@m78/utils";
import { createEvent } from "@m78/hooks";
import isEqual from "lodash/isEqual.js";

export const _DEFAULT_GROUP = "M78-DND-DEFAULT-GROUP";

/** 所有分组数据 */
export const _groupMap: { [key: string]: _GroupState } = {
  [_DEFAULT_GROUP]: {
    scrollParents: [],
    dndMap: {},
  },
};

/** 用于为dnd标记层级关系的context */
export const _levelContext = createContext<_LevelContext>({
  level: 0,
  isDefault: true,
});

/** 在此比例内的区域视为边缘 */
export const EDGE_RATIO = 0.24;

export const _defaultDNDStatus: DNDStatus = {
  dragging: false,
  over: false,
  regular: true,
  left: false,
  right: false,
  top: false,
  bottom: false,
  center: false,
};

export const _defaultDNDEnableInfos: DNDEnableInfos = {
  enable: true,
  all: true,
  left: true,
  right: true,
  top: true,
  bottom: true,
  center: true,
};

export function _useGroup(groupId?: string) {
  return useMemo(() => {
    if (!groupId) {
      return _groupMap[_DEFAULT_GROUP];
    }
    if (!_groupMap[groupId]) {
      _groupMap[groupId] = {
        scrollParents: [],
        dndMap: {},
      };
    }
    return _groupMap[groupId];
  }, [groupId]);
}

/** 判断x, y 是否在指定的DOMRect区间中 */
export function _isBetweenBound(
  { left, top, right, bottom }: Bound,
  x: number,
  y: number
) {
  return x > left && x < right && y > top && y < bottom;
}

/**
 * 通知所有dnd进行状态重置, 应跳过ignoreIds指定的节点, 且状态有变时才进行重置, 否则会造成高频更新,
 * 传入skipEnableReset时, 跳过enables状态的重置
 * */
export const _resetEvent =
  createEvent<(ignoreIds?: string[], skipEnableReset?: boolean) => void>();

/** 通知所有dnd同步位置尺寸信息 */
export const _updateEvent =
  createEvent<(useThrottle: boolean, groupId?: string) => void>();

export const _allValueIsTrue = (obj: AnyObject) => {
  return Object.values(obj).every((v) => v === true);
};

export const _someValueIsTrue = (obj: AnyObject) => {
  return Object.values(obj).some((v) => v === true);
};

export const _getObjectByNewValues = (obj: AnyObject, value: boolean) => {
  const newObj: AnyObject = {};
  Object.keys(obj).forEach((key) => {
    newObj[key] = value;
  });
  return newObj;
};

/** 根据enableDrop获取DNDEnableInfos */
export const _enableDropProcess = (
  enableDrop: DNDProps["enableDrop"],
  current: DNDNode,
  source: DNDNode
): DNDEnableInfos => {
  const enable = isFunction(enableDrop)
    ? enableDrop({ current, source })
    : enableDrop;

  let enables = {
    ..._getObjectByNewValues(_defaultDNDEnableInfos, false),
  } as DNDEnableInfos;

  if (isBoolean(enable)) {
    if (enable) {
      return {
        left: true,
        top: true,
        right: true,
        bottom: true,
        center: true,
        enable: true,
        all: true,
      };
    } else {
      return enables;
    }
  } else if (isObject(enable)) {
    enables = {
      ...enables,
      ...enable,
    };
  }

  return {
    ...enables,
    enable: _someValueIsTrue(enables),
    all: _allValueIsTrue(enables),
  };
};

/** 处理并获取DNDStatus */
export const _statusProcess = (
  dnd: _DNDMapEntry,
  enables: DNDEnableInfos,
  x: number,
  y: number
) => {
  const posEnableInfo = _calcOverStatus(dnd, x, y);

  const status: DNDStatus = {
    dragging: false,
    over: true,
    regular: false,
    top: posEnableInfo.top && enables.top,
    bottom: posEnableInfo.bottom && enables.bottom,
    left: posEnableInfo.left && enables.left,
    right: posEnableInfo.right && enables.right,
    center: posEnableInfo.center && enables.center,
  };

  status.over =
    status.top || status.bottom || status.left || status.right || status.center;

  if (!status.over) {
    status.regular = true;
  }

  return status;
};

/** 计算元光标和指定元素的覆盖状态, 此函数假设光标已在bound范围内 */
export function _calcOverStatus(
  bound: Bound,
  x: number,
  y: number
): DNDPosition {
  const { left, top, right, bottom } = bound;

  // 尺寸
  const width = right - left;
  const height = bottom - top;

  // 触发边缘放置的偏移距离
  const triggerXOffset = width * EDGE_RATIO;
  const triggerYOffset = height * EDGE_RATIO;

  // 各方向上的拖动状态
  const dragTop = y < top + triggerYOffset;
  const dragBottom = !dragTop && y > bottom - triggerYOffset;

  const nextShouldPass = !dragTop && !dragBottom;

  const dragRight = nextShouldPass && x > right - triggerXOffset;
  const dragLeft = nextShouldPass && x < left + triggerXOffset;
  const dragCenter = nextShouldPass && !dragRight && !dragLeft;

  return {
    top: dragTop,
    bottom: dragBottom,
    left: dragLeft,
    right: dragRight,
    center: dragCenter,
  };
}

/** 根据启用和放置状态判定是否可触发Accept */
export function _checkIfAcceptable(
  enables: DNDEnableInfos,
  status: DNDStatus
): boolean {
  if (status.regular) return false;

  if (enables.all && status.over) {
    return true;
  }

  if (enables.left && status.left) return true;
  if (enables.right && status.right) return true;
  if (enables.top && status.top) return true;
  if (enables.bottom && status.bottom) return true;
  return enables.center && status.center;
}

/**
 * 过滤并处理所有在光标区域内的可用dnd节点, 并主要进行以下操作:
 * - 筛选可用并覆盖了光标的dnd
 * - 在初次点击时, 更新所有可见dnd的enable状态
 * - 为满足条件的dnd生成各位置的status
 * */
export function _filterInBoundDNDs(
  ctx: _Context,
  first: boolean,
  xy: TupleNumber
) {
  // 所有被光标命中且进过启用检测的节点
  const inBoundList: _PendingItem[] = [];

  for (const [id, dnd] of Object.entries(ctx.group.dndMap)) {
    if (id === ctx.id) continue;

    // 不可见
    if (!dnd.visible) continue;

    const { enableDrop } = dnd.props;

    const current = dnd.ctx.node;

    const enables = _enableDropProcess(enableDrop, current, ctx.node);

    // 在确认拖动节点后设置到对应dnd状态
    if (first && !isEqual(dnd.ctx.state.enables, enables)) {
      dnd.ctx.setState({
        enables,
      });
    }

    if (!enables.enable) continue;

    // 光标是否在目标区域内
    const inBound = _isBetweenBound(dnd, xy[0], xy[1]);

    // 未在目标区域内
    if (!inBound) continue;

    const status = _statusProcess(dnd, enables, xy[0], xy[1]);

    inBoundList.push({
      dnd,
      enables,
      status,
    });
  }

  return inBoundList;
}

/**
 * 从一组同时命中的dnd中按照指定规则取出一个作为命中点
 * - 规则: 获取level最高的放置点, 若依然存在多个, 获取挂载时间最靠后的一个
 * */
export function _getCurrentTriggerByMultipleTrigger(
  inBoundList: _PendingItem[]
) {
  let current: _PendingItem;

  if (inBoundList.length === 1) {
    current = inBoundList[0];
  } else {
    const list: _PendingItem[] = [];
    let max = 0;

    // 获取level最高的放置点
    for (const item of inBoundList) {
      const level = item.dnd.ctx.level;
      if (level > max) {
        list.splice(0, list.length);
        max = level;
        list.push(item);
        continue;
      }

      if (level === max) {
        list.push(item);
      }
    }

    // 若依然存在多个, 获取挂载时间最靠后的一个
    if (list.length > 1) {
      current = list.reduce((prev, curr) => {
        if (curr.dnd.ctx.mountTime > prev.dnd.ctx.mountTime) return curr;
        return prev;
      }, list[0]);
    } else {
      current = list[0];
    }
  }

  return current;
}

/** 在距离边缘此比例时即开始滚动(比例) */
const AUTO_SCROLL_OFFSET = 0.1;

/** 根据AUTO_SCROLL_OFFSET计算后的距离不大于此值(px) */
const AUTO_SCROLL_OFFSET_MAX = 40;

/** 自动滚动的基准距离, 越大则滚动速度越快 */
const AUTO_SCROLL_BASE_OFFSET = 16;

/**
 * 计算光标在某个元素四个方向的自定滚动比例 0-1,
 * 不包含滚动条的方向返回值始终为0,
 * 元素不包含滚动条时无返回,
 * 同时只会有一个方向有值
 * */
export function _getAutoScrollStatus(
  el: HTMLElement,
  x: number,
  y: number,
  checkOverflowAttr = true
) {
  const isDocOrBody = el === document.documentElement || el === document.body;

  const si = hasScroll(el, checkOverflowAttr);

  if (!isDocOrBody && !si.x && !si.y) return;

  // 滚动容器为body或html根时, 取窗口尺寸
  // eslint-disable-next-line prefer-const
  let { left, top, right, bottom, width, height } = isDocOrBody
    ? {
        left: 0,
        top: 0,
        bottom: window.innerHeight,
        right: window.innerWidth,
        width: window.innerWidth,
        height: window.innerHeight,
      }
    : el.getBoundingClientRect();

  // 取最小、最大触发位置
  left = Math.max(left, 0);
  top = Math.max(top, 0);
  right = Math.min(right, window.innerWidth);
  bottom = Math.min(bottom, window.innerHeight);

  const xTriggerOffset = Math.min(
    AUTO_SCROLL_OFFSET_MAX,
    width * AUTO_SCROLL_OFFSET
  );
  const yTriggerOffset = Math.min(
    AUTO_SCROLL_OFFSET_MAX,
    height * AUTO_SCROLL_OFFSET
  );

  // 计算偏移
  left += xTriggerOffset;
  top += yTriggerOffset;
  right -= xTriggerOffset;
  bottom -= yTriggerOffset;

  let t = 0;
  let r = 0;
  let b = 0;
  let l = 0;

  // 在y轴范围内, 且在可触发自动滚动的范围内(边界前后 yTriggerOffset 范围)
  if (x > left && x < right) {
    if (y < top && y > top - yTriggerOffset * 2) {
      t = Math.min(1, (top - y) / yTriggerOffset);
    }

    if (y > bottom && y < bottom + yTriggerOffset * 2) {
      b = Math.min(1, (y - bottom) / yTriggerOffset);
    }
  }

  // 在x轴范围内
  if (y > top && y < bottom) {
    if (x < left && x > left - xTriggerOffset * 2) {
      l = Math.min(1, (left - x) / xTriggerOffset);
    }

    if (x > right && x < right + xTriggerOffset * 2) {
      r = Math.min(1, (x - right) / xTriggerOffset);
    }
  }

  return {
    top: si.y ? t : 0,
    bottom: si.y ? b : 0,
    left: si.x ? l : 0,
    right: si.x ? r : 0,
  };
}

/**
 * 根据getAutoScrollStatus的返回值滚动元素
 *
 * @param element 滚动元素
 * @param enable status有效时, 是否启用自动滚动, 若滚动一开始, 至少需要传入一次false来关闭自动滚动
 * @param status 光标在模板元素边缘的信息, getAutoScrollStatus的返回值
 * */
export function _autoScrollByStatus(
  element: HTMLElement,
  enable: boolean,
  status: ReturnType<typeof _getAutoScrollStatus>
) {
  const el = element as HTMLElement & { ctx: _AutoScrollCtx };

  // 滚动元素本身是一个非常理想的存储局部滚动状态的对象
  if (!el.ctx) {
    el.ctx = {} as _AutoScrollCtx;
  }

  el.ctx.autoScrollDown = enable;

  if (!el || !status) return;

  // 基础滚动距离
  el.ctx.autoScrollVal = 1;

  if (status.bottom) {
    el.ctx.autoScrollPosKey = "scrollTop";
    el.ctx.autoScrollType = 1;
    el.ctx.autoScrollVal += status.bottom * AUTO_SCROLL_BASE_OFFSET;
  }

  if (status.left) {
    el.ctx.autoScrollPosKey = "scrollLeft";
    el.ctx.autoScrollType = 2;
    el.ctx.autoScrollVal += status.left * AUTO_SCROLL_BASE_OFFSET;
  }

  if (status.top) {
    el.ctx.autoScrollPosKey = "scrollTop";
    el.ctx.autoScrollType = 2;
    el.ctx.autoScrollVal += status.top * AUTO_SCROLL_BASE_OFFSET;
  }

  if (status.right) {
    el.ctx.autoScrollPosKey = "scrollLeft";
    el.ctx.autoScrollType = 1;
    el.ctx.autoScrollVal += status.right * AUTO_SCROLL_BASE_OFFSET;
  }

  // 根据状态开关滚动动画
  if (!(status.bottom || status.top || status.left || status.right)) {
    el.ctx.autoScrollToggle = false;
  } else {
    if (!el.ctx.autoScrollToggle) {
      autoScroll(el);
    }
    el.ctx.autoScrollToggle = true;
  }
}

/** 根据当前的AutoScrollCtx来自动滚动目标元素 */
function autoScroll(el: HTMLElement & { ctx: _AutoScrollCtx }) {
  raf(() => {
    if (el.ctx.autoScrollType === 1) {
      el[el.ctx.autoScrollPosKey] += el.ctx.autoScrollVal;

      // 处理浏览器兼容
      if (el === document.documentElement) {
        document.body[el.ctx.autoScrollPosKey] += el.ctx.autoScrollVal;
      }
    } else {
      el[el.ctx.autoScrollPosKey] -= el.ctx.autoScrollVal;

      // 处理浏览器兼容
      if (el === document.documentElement) {
        document.body[el.ctx.autoScrollPosKey] -= el.ctx.autoScrollVal;
      }
    }

    if (!el.ctx.autoScrollDown || !el.ctx.autoScrollToggle) return;

    autoScroll(el);
  });
}

/** 禁止拖动的元素tagName */
const ignoreReg = /^(INPUT|TEXTAREA|BUTTON|SELECT|AUDIO|VIDEO)$/;

/**  根据事件元素类型决定是否禁止拖动 */
export function _isIgnoreEl(event?: any, ignoreElFilter?: DNDProps["ignore"]) {
  const el = event?.target;

  if (!el) return false;

  const tagName = el.tagName || "";

  if (ignoreReg.test(tagName)) return true;

  const editable = el.getAttribute && el.getAttribute("contenteditable");

  if (editable) return true;

  if (ignoreElFilter) {
    return ignoreElFilter(el);
  }

  return false;
}
