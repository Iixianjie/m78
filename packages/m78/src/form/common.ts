import {
  ensureArray,
  isArray,
  isEmpty,
  isNumber,
  isObject,
  NamePath,
  stringifyNamePath,
} from "@m78/utils";
import { _Context, _State, FormNamesNotify } from "./types.js";
import isEqual from "lodash/isEqual.js";

/** 获取指定name的state, 状态对象还不存在时会自动进行创建 */
export function _getState(ctx: _Context, name: NamePath) {
  const nameKey = stringifyNamePath(name);
  let st = ctx.state[nameKey];

  if (!st) {
    st = {
      name,
      errors: [],
    };
    ctx.state[nameKey] = st;
  }

  return st;
}

/** 遍历已存在的所有state */
export function _eachState(ctx: _Context, cb: (st: _State) => void) {
  for (const key of Object.keys(ctx.state)) {
    cb(ctx.state[key] || {});
  }
}

/** 根据values递归为其设置或初始化state */

/**
 * 用于创建update/change事件回调的过滤器, 只在name对应的值发生变化时才会触发回调,
 * 若传入deps, deps中指定的值发生变化时也会触发回调
 * */
export function _notifyFilter(
  name: NamePath,
  notify: FormNamesNotify,
  deps: NamePath[] = []
): FormNamesNotify {
  // 触发规则, 值更新时, 其所有上级/下级都会触发

  const np = ensureArray(name);

  return (triggerName, relation) => {
    if (!triggerName) {
      notify(triggerName, relation);
      return;
    }

    const np2 = ensureArray(triggerName);
    const names = [np, ...deps.map(ensureArray)];

    if (relation) {
      for (const n of names) {
        // 需要更新关联值
        const min = Math.min(n.length, np2.length);

        const arr1 = n.slice(0, min);
        const arr2 = np2.slice(0, min);

        if (isEqual(arr1, arr2)) {
          notify(triggerName, relation);
          break;
        }
      }
    } else {
      for (const n of names) {
        // 不关联, 直接对比
        if (isEqual(n, np2)) {
          notify(triggerName, relation);
          break;
        }
      }
    }
  };
}

/** 用于namePath路径的通配占位, 匹配eachSchema等没有name的层级 */
export const _ANY_NAME_PLACE_HOLD = "__ANY_NAME_PLACE_HOLD__";

/** 用于在某些情况作为根schema的name标注 */
export const ROOT_SCHEMA_NAME = "__ROOT_SCHEMA_NAME__";

/** 递归删除指定的namePath值, 支持在namePath中使用ANY_NAME_PLACE_HOLD进行通配占位 */
export function _recursionDeleteNamePath(values: any, names: NamePath) {
  const name = ensureArray(names);

  if (!name.length || isEmpty(values)) return;

  const [currentName, ...rest] = name;

  const isArr = isArray(values);
  const isObj = isObject(values);

  if (!isArr && !isObj) return;

  if (name.length === 1) {
    if (isObj) {
      // 清理全部
      if (currentName === _ANY_NAME_PLACE_HOLD) {
        Object.keys(values).forEach((key) => delete values[key]);
        return;
      }
      delete values[currentName];
    }
    if (isArr) {
      // 清理全部
      if (currentName === _ANY_NAME_PLACE_HOLD) {
        values.splice(0, values.length);
        return;
      }

      if (isNumber(currentName)) {
        values.splice(currentName, 1);
      }
    }
    return;
  }

  if (currentName !== _ANY_NAME_PLACE_HOLD) {
    _recursionDeleteNamePath(values[currentName], rest);
    return;
  }

  if (isObj) {
    Object.keys(values).forEach((key) => {
      _recursionDeleteNamePath(values[key], rest);
    });
  }

  if (isArr) {
    values.forEach((v) => {
      _recursionDeleteNamePath(v, rest);
    });
  }
}
