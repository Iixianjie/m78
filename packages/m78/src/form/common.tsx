import { isObject } from "@m78/utils";
import { FormAdaptor } from "./types.js";

/** 默认的表单 onChange 值获取器 */
export const _defaultValueGetter = (value: any) => {
  if (isObject(value) && "target" in value && "value" in value.target) {
    return value.target.value;
  }
  return value;
};

/** 默认适配器 */
export const _defaultAdaptor: FormAdaptor = (args) => {
  if (!args.element) return null;

  return args.binder(args.element, {
    ...args.element.props,
    ...args.bind,
  });
};

export const EMPTY_NAME = "__EMPTY_NAME__";
export const EMPTY_LIST_NAME = "__EMPTY_LIST_NAME__";
