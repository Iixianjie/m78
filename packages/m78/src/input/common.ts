export const _NAMESPACE = "Input";
