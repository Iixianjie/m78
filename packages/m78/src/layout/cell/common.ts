import { MediaQueryMeta, MediaQueryObject } from "../media-query/types.js";
import { _mediaQueryGetter } from "../media-query/common.js";
import {
  CellColMediaQueryProps,
  CellColNumberOrMediaQueryProps,
  CellColProps,
} from "./types";
import { isNumber, isObject, isTruthyOrZero } from "@m78/utils";

/**
 * 根据MediaQueryMeta从GridColProps中获取GridColMediaQueryProps
 * 获取遵循以下规则:
 * - 当前媒体尺寸不包含有效配置时，从小于此区间的媒体尺寸中依次获取并拿到第一个有效配置
 * - 当任何区间都不能获取到有效值时，获取直接传入的配置
 * - 如下所示
 * - xxl.col > md.col > xs.col > props.col
 * - xxl.offset > md.offset > xs.offset > props.offset
 * */
export function _getCurrentMqProps(
  mqMeta: MediaQueryMeta,
  {
    col,
    offset,
    move,
    order,
    flex,
    hidden,
    align,
    xs,
    sm,
    md,
    lg,
    xl,
    xxl,
    className,
    style,
  }: CellColProps
) {
  const mqObject: MediaQueryObject<CellColNumberOrMediaQueryProps> = {
    xs,
    sm,
    md,
    lg,
    xl,
    xxl,
  };

  const obj = {
    col: _mediaQueryGetter(
      mqMeta!,
      mqObject,
      col,
      (item) => isNumber(item) || isNumber(item?.col)
    ),
    offset: _mediaQueryGetter(
      mqMeta!,
      mqObject,
      offset,
      (item) => !isNumber(item) && isNumber(item?.offset)
    ),
    move: _mediaQueryGetter(
      mqMeta!,
      mqObject,
      move,
      (item) => !isNumber(item) && isNumber(item?.move)
    ),
    order: _mediaQueryGetter(
      mqMeta!,
      mqObject,
      order,
      (item) => !isNumber(item) && isNumber(item?.order)
    ),
    flex: _mediaQueryGetter(
      mqMeta!,
      mqObject,
      flex as any,
      (item) => !isNumber(item) && isTruthyOrZero(item?.flex)
    ),
    hidden: _mediaQueryGetter(
      mqMeta!,
      mqObject,
      hidden as any,
      (item) => !isNumber(item) && item?.hidden,
      true
    ),
    align: _mediaQueryGetter(
      mqMeta!,
      mqObject,
      align as any,
      (item) => !isNumber(item) && item?.align
    ),
    className: _mediaQueryGetter(
      mqMeta!,
      mqObject,
      className as any,
      (item) => !isNumber(item) && item?.className
    ),
    style: _mediaQueryGetter(
      mqMeta!,
      mqObject,
      style as any,
      (item) => !isNumber(item) && item?.style
    ),
  };

  const gridColMediaQueryProps: any = {};

  Object.entries(obj).forEach(([k, v]) => {
    if (isObject(v)) {
      gridColMediaQueryProps[k] = v[k];
    } else {
      gridColMediaQueryProps[k] = v;
    }
  });

  return gridColMediaQueryProps as CellColMediaQueryProps;
}
