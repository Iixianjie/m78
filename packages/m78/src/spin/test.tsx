import renderer from "react-test-renderer";
import React from "react";
import { FullSize } from "../common";
import { Spin } from "./index";

test("spin base", () => {
  const r = renderer
    .create(
      <div>
        <Spin size={FullSize.small} />
        <Spin className="ml-12" />
        <Spin size={FullSize.large} className="ml-12" />
        <Spin size={FullSize.big} className="ml-12" />
        <Spin size={FullSize.small} text="加载中" />

        <div className="mt-12">
          <Spin size={FullSize.small} inline text="加载中" />
          <Spin
            size={FullSize.small}
            inline
            text="自定义文本"
            className="ml-12"
          />
        </div>

        <div className="d-spin-full-wrap">
          Lorem ipsum
          <Spin open full />
        </div>
      </div>
    )
    .toJSON();

  expect(r).toMatchSnapshot();
});
