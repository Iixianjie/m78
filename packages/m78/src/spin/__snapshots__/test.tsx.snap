// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spin base 1`] = `
<div>
  <div
    className="m78 m78-spin __small"
    style={
      Object {
        "__progress": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "visibility": "hidden",
      }
    }
  >
    <div
      className="m78-spin_inner"
      style={
        Object {
          "top": undefined,
        }
      }
    >
      <svg
        className="m78-spin-svg m78-spin_unit"
        fill="transparent"
        height={200}
        viewBox="0 0 200 200"
        width={200}
      >
        <circle
          className="m78-spin-svg_bg"
          cx={100}
          cy={100}
          r={70}
          strokeWidth={60}
        />
        <circle
          cx={100}
          cy={100}
          r={70}
          stroke="url(#Gradient1)"
          strokeDasharray="628.3185307179587 628.3185307179587"
          strokeLinecap="round"
          strokeWidth={60}
        >
          <animate
            attributeName="stroke-dashoffset"
            dur="5s"
            repeatCount="indefinite"
            values="628.3185307179587;0;628.3185307179587"
          />
        </circle>
        <defs>
          <linearGradient
            id="Gradient1"
          >
            <stop
              className="m78-spin-svg_stop1"
              offset="0%"
            />
            <stop
              className="m78-spin-svg_stop2"
              offset="100%"
            />
          </linearGradient>
        </defs>
      </svg>
    </div>
  </div>
  <div
    className="ml-12 m78 m78-spin"
    style={
      Object {
        "__progress": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "visibility": "hidden",
      }
    }
  >
    <div
      className="m78-spin_inner"
      style={
        Object {
          "top": undefined,
        }
      }
    >
      <svg
        className="m78-spin-svg m78-spin_unit"
        fill="transparent"
        height={200}
        viewBox="0 0 200 200"
        width={200}
      >
        <circle
          className="m78-spin-svg_bg"
          cx={100}
          cy={100}
          r={70}
          strokeWidth={60}
        />
        <circle
          cx={100}
          cy={100}
          r={70}
          stroke="url(#Gradient1)"
          strokeDasharray="628.3185307179587 628.3185307179587"
          strokeLinecap="round"
          strokeWidth={60}
        >
          <animate
            attributeName="stroke-dashoffset"
            dur="5s"
            repeatCount="indefinite"
            values="628.3185307179587;0;628.3185307179587"
          />
        </circle>
        <defs>
          <linearGradient
            id="Gradient1"
          >
            <stop
              className="m78-spin-svg_stop1"
              offset="0%"
            />
            <stop
              className="m78-spin-svg_stop2"
              offset="100%"
            />
          </linearGradient>
        </defs>
      </svg>
    </div>
  </div>
  <div
    className="ml-12 m78 m78-spin __large"
    style={
      Object {
        "__progress": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "visibility": "hidden",
      }
    }
  >
    <div
      className="m78-spin_inner"
      style={
        Object {
          "top": undefined,
        }
      }
    >
      <svg
        className="m78-spin-svg m78-spin_unit"
        fill="transparent"
        height={200}
        viewBox="0 0 200 200"
        width={200}
      >
        <circle
          className="m78-spin-svg_bg"
          cx={100}
          cy={100}
          r={70}
          strokeWidth={60}
        />
        <circle
          cx={100}
          cy={100}
          r={70}
          stroke="url(#Gradient1)"
          strokeDasharray="628.3185307179587 628.3185307179587"
          strokeLinecap="round"
          strokeWidth={60}
        >
          <animate
            attributeName="stroke-dashoffset"
            dur="5s"
            repeatCount="indefinite"
            values="628.3185307179587;0;628.3185307179587"
          />
        </circle>
        <defs>
          <linearGradient
            id="Gradient1"
          >
            <stop
              className="m78-spin-svg_stop1"
              offset="0%"
            />
            <stop
              className="m78-spin-svg_stop2"
              offset="100%"
            />
          </linearGradient>
        </defs>
      </svg>
    </div>
  </div>
  <div
    className="ml-12 m78 m78-spin __big"
    style={
      Object {
        "__progress": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "visibility": "hidden",
      }
    }
  >
    <div
      className="m78-spin_inner"
      style={
        Object {
          "top": undefined,
        }
      }
    >
      <svg
        className="m78-spin-svg m78-spin_unit"
        fill="transparent"
        height={200}
        viewBox="0 0 200 200"
        width={200}
      >
        <circle
          className="m78-spin-svg_bg"
          cx={100}
          cy={100}
          r={70}
          strokeWidth={60}
        />
        <circle
          cx={100}
          cy={100}
          r={70}
          stroke="url(#Gradient1)"
          strokeDasharray="628.3185307179587 628.3185307179587"
          strokeLinecap="round"
          strokeWidth={60}
        >
          <animate
            attributeName="stroke-dashoffset"
            dur="5s"
            repeatCount="indefinite"
            values="628.3185307179587;0;628.3185307179587"
          />
        </circle>
        <defs>
          <linearGradient
            id="Gradient1"
          >
            <stop
              className="m78-spin-svg_stop1"
              offset="0%"
            />
            <stop
              className="m78-spin-svg_stop2"
              offset="100%"
            />
          </linearGradient>
        </defs>
      </svg>
    </div>
  </div>
  <div
    className="m78 m78-spin __small"
    style={
      Object {
        "__progress": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "visibility": "hidden",
      }
    }
  >
    <div
      className="m78-spin_inner"
      style={
        Object {
          "top": undefined,
        }
      }
    >
      <svg
        className="m78-spin-svg m78-spin_unit"
        fill="transparent"
        height={200}
        viewBox="0 0 200 200"
        width={200}
      >
        <circle
          className="m78-spin-svg_bg"
          cx={100}
          cy={100}
          r={70}
          strokeWidth={60}
        />
        <circle
          cx={100}
          cy={100}
          r={70}
          stroke="url(#Gradient1)"
          strokeDasharray="628.3185307179587 628.3185307179587"
          strokeLinecap="round"
          strokeWidth={60}
        >
          <animate
            attributeName="stroke-dashoffset"
            dur="5s"
            repeatCount="indefinite"
            values="628.3185307179587;0;628.3185307179587"
          />
        </circle>
        <defs>
          <linearGradient
            id="Gradient1"
          >
            <stop
              className="m78-spin-svg_stop1"
              offset="0%"
            />
            <stop
              className="m78-spin-svg_stop2"
              offset="100%"
            />
          </linearGradient>
        </defs>
      </svg>
      <span
        className="m78-spin_text"
      >
        加载中
        <span
          className="m78-spin_ellipsis"
        />
      </span>
    </div>
  </div>
  <div
    className="mt-12"
  >
    <div
      className="m78 m78-spin __small __inline"
      style={
        Object {
          "__progress": 0,
          "opacity": 0,
          "pointerEvents": "none",
          "visibility": "hidden",
        }
      }
    >
      <div
        className="m78-spin_inner"
        style={
          Object {
            "top": undefined,
          }
        }
      >
        <svg
          className="m78-spin-svg m78-spin_unit"
          fill="transparent"
          height={200}
          viewBox="0 0 200 200"
          width={200}
        >
          <circle
            className="m78-spin-svg_bg"
            cx={100}
            cy={100}
            r={70}
            strokeWidth={60}
          />
          <circle
            cx={100}
            cy={100}
            r={70}
            stroke="url(#Gradient1)"
            strokeDasharray="628.3185307179587 628.3185307179587"
            strokeLinecap="round"
            strokeWidth={60}
          >
            <animate
              attributeName="stroke-dashoffset"
              dur="5s"
              repeatCount="indefinite"
              values="628.3185307179587;0;628.3185307179587"
            />
          </circle>
          <defs>
            <linearGradient
              id="Gradient1"
            >
              <stop
                className="m78-spin-svg_stop1"
                offset="0%"
              />
              <stop
                className="m78-spin-svg_stop2"
                offset="100%"
              />
            </linearGradient>
          </defs>
        </svg>
        <span
          className="m78-spin_text"
        >
          加载中
          <span
            className="m78-spin_ellipsis"
          />
        </span>
      </div>
    </div>
    <div
      className="ml-12 m78 m78-spin __small __inline"
      style={
        Object {
          "__progress": 0,
          "opacity": 0,
          "pointerEvents": "none",
          "visibility": "hidden",
        }
      }
    >
      <div
        className="m78-spin_inner"
        style={
          Object {
            "top": undefined,
          }
        }
      >
        <svg
          className="m78-spin-svg m78-spin_unit"
          fill="transparent"
          height={200}
          viewBox="0 0 200 200"
          width={200}
        >
          <circle
            className="m78-spin-svg_bg"
            cx={100}
            cy={100}
            r={70}
            strokeWidth={60}
          />
          <circle
            cx={100}
            cy={100}
            r={70}
            stroke="url(#Gradient1)"
            strokeDasharray="628.3185307179587 628.3185307179587"
            strokeLinecap="round"
            strokeWidth={60}
          >
            <animate
              attributeName="stroke-dashoffset"
              dur="5s"
              repeatCount="indefinite"
              values="628.3185307179587;0;628.3185307179587"
            />
          </circle>
          <defs>
            <linearGradient
              id="Gradient1"
            >
              <stop
                className="m78-spin-svg_stop1"
                offset="0%"
              />
              <stop
                className="m78-spin-svg_stop2"
                offset="100%"
              />
            </linearGradient>
          </defs>
        </svg>
        <span
          className="m78-spin_text"
        >
          自定义文本
          <span
            className="m78-spin_ellipsis"
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="d-spin-full-wrap"
  >
    Lorem ipsum
    <div
      className="m78 m78-spin __full"
      style={
        Object {
          "__progress": 0,
          "opacity": 0,
          "pointerEvents": "none",
          "visibility": "hidden",
        }
      }
    >
      <div
        className="m78-spin_inner"
        style={
          Object {
            "top": undefined,
          }
        }
      >
        <svg
          className="m78-spin-svg m78-spin_unit"
          fill="transparent"
          height={200}
          viewBox="0 0 200 200"
          width={200}
        >
          <circle
            className="m78-spin-svg_bg"
            cx={100}
            cy={100}
            r={70}
            strokeWidth={60}
          />
          <circle
            cx={100}
            cy={100}
            r={70}
            stroke="url(#Gradient1)"
            strokeDasharray="628.3185307179587 628.3185307179587"
            strokeLinecap="round"
            strokeWidth={60}
          >
            <animate
              attributeName="stroke-dashoffset"
              dur="5s"
              repeatCount="indefinite"
              values="628.3185307179587;0;628.3185307179587"
            />
          </circle>
          <defs>
            <linearGradient
              id="Gradient1"
            >
              <stop
                className="m78-spin-svg_stop1"
                offset="0%"
              />
              <stop
                className="m78-spin-svg_stop2"
                offset="100%"
              />
            </linearGradient>
          </defs>
        </svg>
      </div>
    </div>
  </div>
</div>
`;
